src/unix/event.rs

```diff
@@ -0,0 +1,186 @@
+
+/// EventNotifier
+/// This is a generic event notifier that can be used with eventfd or pipefd.
+/// It allows writing a value to the file descriptor to notify an event.
+///
+/// # Examples
+///
+/// ```
+/// use std::os::fd::FromRawFd;
+/// use std::os::unix::io::IntoRawFd;
+/// use vmm_sys_util::event::EventNotifier;
+/// let (_, writer) = std::io::pipe().expect("Failed to create pipe");
+/// let notifier = unsafe { EventNotifier::from_raw_fd(writer.into_raw_fd()) };
+/// ```
+#[derive(Debug)]
+pub struct EventNotifier {
+    fd: File,
+}
+
+impl EventNotifier {
+    /// Write a value to the EventNotifier's fd
+    /// Writing 1 to fd is for compatibility with Eventfd
+    pub fn notify(&self) -> result::Result<(), io::Error> {
+        let v = 1u64;
+        (&self.fd).write_all(&v.to_ne_bytes())
+    }
+
+    /// Clone this EventNotifier.
+    pub fn try_clone(&self) -> result::Result<EventNotifier, io::Error> {
+        Ok(EventNotifier {
+            fd: self.fd.try_clone()?,
+        })
+    }
+}
+
+impl AsRawFd for EventNotifier {
+    fn as_raw_fd(&self) -> std::os::unix::prelude::RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+/// Convert a raw file descriptor into an EventNotifier.
+impl FromRawFd for EventNotifier {
+    unsafe fn from_raw_fd(fd: std::os::unix::prelude::RawFd) -> Self {
+        EventNotifier {
+            fd: File::from_raw_fd(fd),
+        }
+    }
+}
+
+/// EventReceiver
+/// This is a generic event consumer that can be used with eventfd or pipefd.
+/// It allows reading a value from the file descriptor to consume an event.
+///
+/// # Examples
+///
+/// ```
+/// use std::os::fd::FromRawFd;
+/// use std::os::unix::io::IntoRawFd;
+/// use vmm_sys_util::event::EventConsumer;
+/// let (reader, _) = std::io::pipe().expect("Failed to create pipe");
+/// let consumer = unsafe { EventConsumer::from_raw_fd(reader.into_raw_fd()) };
+/// ```
+#[derive(Debug)]
+pub struct EventConsumer {
+    fd: File,
+}
+
+impl EventConsumer {
+    /// Read a value from the EventConsumer.
+    pub fn consume(&self) -> result::Result<(), io::Error> {
+        let mut buf = [0u8; size_of::<u64>()];
+        (&self.fd).read_exact(buf.as_mut_slice()).map(|_| Ok(()))?
+    }
+
+    /// Clone this EventConsumer.
+    pub fn try_clone(&self) -> result::Result<EventConsumer, io::Error> {
+        Ok(EventConsumer {
+            fd: self.fd.try_clone()?,
+        })
+    }
+}
+
+impl AsRawFd for EventConsumer {
+    fn as_raw_fd(&self) -> std::os::unix::prelude::RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+/// Convert a raw file descriptor into an EventConsumer.
+impl FromRawFd for EventConsumer {
+    unsafe fn from_raw_fd(fd: std::os::unix::prelude::RawFd) -> Self {
+        EventConsumer {
+            fd: File::from_raw_fd(fd),
+        }
+    }
+}
+
+/// Create a new EventNotifier and EventConsumer using a pipe.
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
+pub fn new_event_notifier_and_consumer(
+    flag: i32,
```

Comment:
We should document what flag is supposed to be.
Are we sure it's the same for `fcntl` and for `EventFd::new()` ?

Reply:
Yeah, something like that! Although like this it only allows to set one flag at a time, instead of a combination of them (sorry, been working in C too much recently that I apparently forgot how Rust enums work :sob:), so we'll need something like [`bitflags!`](https://docs.rs/bitflags/latest/bitflags/) (to clarify, the value of each flag should be a synthetic value introduced by this crate, e.g. NonBlock = 0x1, CloseOnExec = 0x2, and then `From::from` checks which synthetic bits are set and converts it to the platform specific combination of bits that correspond to them)

Reply:
Sorry if I'm late here, but I prefer a builder type of API, that can handle both O_CLOEXEC and NONBLOCK since both are common to pipes and eventfd, something like:
```rust
let (en, ec) = EventBuilder::new()
    .set_cloexec()
    .set_nonblock()
    .build()
```

so the flags are optional


for instance we do care about cloexec in virtiofsd

Reply:
I'm not suggesting the names btw, we can also do something similar to what is done inside the std
```rust
let (en, ec) = EventBuilder::new()
    .closeOnExec(true)
    .nonBlock(true)
    .build()
```

or without the bool parameter